fendermint/actors/eam/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::state::State;
+pub use fil_actor_eam::Method;
+use fil_actor_eam::{
+    Create2Params, Create2Return, CreateExternalParams, CreateExternalReturn, CreateParams,
+    CreateReturn, EamActor,
+};
+use fil_actors_runtime::actor_error;
+use fil_actors_runtime::runtime::{ActorCode, Runtime};
+use fil_actors_runtime::{actor_dispatch_unrestricted, ActorError};
+use fvm_shared::address::Address;
+
+mod state;
+
+#[cfg(feature = "fil-actor")]
+fil_actors_runtime::wasm_trampoline!(IPCEamActor);
+
+pub const IPC_EAM_ACTOR_NAME: &str = "EVMAddressManager";
+
+pub struct IPCEamActor;
+
+impl IPCEamActor {
+    pub fn constructor(
+        rt: &impl Runtime,
+        whitelisted_deployers: Vec<Address>,
+    ) -> Result<(), ActorError> {
+        EamActor::constructor(rt)?;
+
+        let mut st = State::new(rt.store())?;
+        for d in whitelisted_deployers {
+            st.add_deployer(rt.store(), &d)?;
+        }
+
+        rt.create(&st)?;
+
+        Ok(())
+    }
+
+    /// Create a new contract per the EVM's CREATE rules.
+    ///
+    /// Permissions: May be called by the EVM.
+    pub fn create(rt: &impl Runtime, params: CreateParams) -> Result<CreateReturn, ActorError> {
+        Self::ensure_deployer_allowed(rt)?;
+        EamActor::create(rt, params)
+    }
+
+    /// Create a new contract per the EVM's CREATE2 rules.
+    ///
+    /// Permissions: May be called by the EVM.
+    pub fn create2(rt: &impl Runtime, params: Create2Params) -> Result<Create2Return, ActorError> {
+        Self::ensure_deployer_allowed(rt)?;
+        EamActor::create2(rt, params)
+    }
+
+    /// Create a new contract from off-chain.
+    ///
+    /// When called by an EthAccount, this method will compute the new actor's address according to
+    /// the `CREATE` rules. When called by a "native" Account, this method will derive the address
+    /// from the _hash_ of the caller's key address.
+    ///
+    /// Permissions: May be called by builtin or eth accounts.
+    pub fn create_external(
+        rt: &impl Runtime,
+        params: CreateExternalParams,
+    ) -> Result<CreateExternalReturn, ActorError> {
+        Self::ensure_deployer_allowed(rt)?;
+        EamActor::create_external(rt, params)
+    }
+
+    fn ensure_deployer_allowed(rt: &impl Runtime) -> Result<(), ActorError> {
+        let caller = rt.message().caller();
+
+        let state: State = rt.state()?;
+        if !state.can_deploy(rt.store(), &caller)? {
+            return Err(ActorError::forbidden(String::from(
+                "sender not allowed to deploy contracts",
+            )));
+        }
+
+        Ok(())
+    }
+}
+
+impl ActorCode for IPCEamActor {
+    type Methods = Method;
+
+    fn name() -> &'static str {
+        IPC_EAM_ACTOR_NAME
+    }
+
+    actor_dispatch_unrestricted! {
+        Constructor => constructor,
+        Create => create,
+        Create2 => create2,
+        CreateExternal => create_external,
+    }
```

Comment:
If any and all methods need to be dispatched, isn't there a way not to have to enumerate them at all?

Reply:
We can create macro rules on top of this. I think this is probably the convention used and the actor is quite simple. I guess it's ok?

Reply:
I mean, there must be some general method that would simply forward, like `handle_message(method_number, params)` ?

Reply:
yeah, there is, but `actor_dispatch_unrestricted` actually contains some checking logic as well.

Reply:
The `ActorCode` trait has this `invoke_method` method that you can trivially implement for this actor, and then you don't need to implement all the individual methods, because they all check the same thing. Something like this:

```
        fn invoke_method<RT>(
            rt: &RT,
            method: fvm_shared::MethodNum,
            args: Option<fvm_ipld_encoding::ipld_block::IpldBlock>,
        ) -> Result<Option<fvm_ipld_encoding::ipld_block::IpldBlock>, ActorError>
        where
            RT: Runtime,
            RT::Blockstore: Clone,
        {
            Self::ensure_deployer_allowed(rt)?;
            EamActor::invoke_method(rt, method, args)
        }
```

Reply:
I don't see any documentation on `actor_dispatch_unrestricted` so can't say what's the difference between that and `actor_dispatch`, but the only check I see it does is the method number. That will be performed by `EamActor::invoke_method` as well.

Reply:
I agree with @aakoshh. Performing the permissioning at the invoke_method entrypoint should be enough (we should leave the Constructor) out of this extra layer of permissioning, since it can only be invoked by the system actor and it already performs that check.

https://github.com/filecoin-project/builtin-actors/blob/master/actors/eam/src/lib.rs#L237

Reply:
This way we can get rid of every method here.

Reply:
Updated accordingly